### PR TITLE
Documents/Kit: Fix - Revisions are not being saved in Site Settings (…

### PR DESCRIPTION
**Pull request by iNewLegend**

### CODE REVIEW

**Comment by iNewLegend** [comment]
TODO: Tests.

File: core/kits/documents/kit.php [hunk 1]
```diff
@@ -73,7 +73,25 @@ public function get_css_wrapper_selector() {
...
+			if ( ! $post_has_changed ) {
+				$last_revision_document = Plugin::instance()->documents->get( $last_revision->ID );
+				$post_document = Plugin::instance()->documents->get( $post->ID );
+
+				$last_revision_settings = $last_revision_document->get_settings();
+				$post_settings = $post_document->get_settings();
+
+				// Its better to add crc32 signature for each revision and then only compare one part of the checksum.
+				return $last_revision_settings !== $post_settings;
+			}
```
Comment (matipojo) [inline core/kits/documents/kit.php:88]: A filter should return by default the first argument he received.

File: core/kits/documents/kit.php [hunk 2]
```diff
@@ -73,7 +73,25 @@ public function get_css_wrapper_selector() {
 		return '.elementor-kit-' . $this->get_main_id();
 	}
 
+	public function handle_revisions_changed() {
```
Comment (matipojo) [inline core/kits/documents/kit.php:76]: Move to base document.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/kits/documents/kit.php [hunk 3]
```diff
@@ -73,7 +73,27 @@ public function get_css_wrapper_selector() {
...
+			// In case default, didn't determine the changes.
+			if ( ! $post_has_changed ) {
+				$last_revision_document = Plugin::instance()->documents->get( $last_revision->ID );
+				$post_document = Plugin::instance()->documents->get( $post->ID );
+
+				$last_revision_settings = $last_revision_document->get_settings();
+				$post_settings = $post_document->get_settings();
+
+				// Its better to add crc32 signature for each revision and then only compare one part of the checksum.
+				return $last_revision_settings !== $post_settings;
```
Comment (matipojo) [inline core/kits/documents/kit.php:87]:
```suggestion
				$post_has_changed = $last_revision_settings !== $post_settings;
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/kits/documents/kit.php (hunk 1 above)
Comment (iNewLegend) [reply core/kits/documents/kit.php:88]: Handled.

File: core/kits/documents/kit.php (hunk 2 above)
Comment (iNewLegend) [reply core/kits/documents/kit.php]: Handled.

File: core/kits/documents/kit.php (hunk 3 above)
Comment (iNewLegend) [reply core/kits/documents/kit.php]: Handled.

File: core/base/document.php [hunk 4]
```diff
@@ -1284,4 +1284,23 @@ protected function get_post_statuses() {
 	protected function get_have_a_look_url() {
 		return $this->get_permalink();
 	}
+
+	public function handle_revisions_changed() {
+		add_filter( 'wp_save_post_revision_post_has_changed', function ( $post_has_changed, $last_revision, $post ) {
```
Comment (matipojo) [inline core/base/document.php:1289]: Convert the function to a method andn remove the filter after the document is saved.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: tests/phpunit/elementor/core/kits/documents/kits/kit.php [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+		$this->assertCount( 1, Revisions_Manager::get_revisions( $kit->get_main_id() ) );
+		$kit = Plugin::$instance->documents->get( $kit->get_id(), false );
+
+		$kit->set_settings( 'post_status', 'published');
+		$kit->save( [ 'settings' => $kit->get_settings() ] );
+
+		$this->assertCount( 2, Revisions_Manager::get_revisions( $kit->get_main_id(), [], false ) );
+
+		$kit->set_settings( 'post_status', 'published');
```
Comment (matipojo) [inline tests/phpunit/elementor/core/kits/documents/kits/kit.php:22]:
```suggestion
		// Assert it not creates a revision if there are no changes.
		$kit->set_settings( 'post_status', 'published');
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: tests/phpunit/elementor/core/kits/documents/kits/kit.php (hunk 5 above)
Comment (iNewLegend) [reply tests/phpunit/elementor/core/kits/documents/kits/kit.php:22]: _elementor_data

File: core/kits/documents/kit.php [hunk 6]
```diff
@@ -74,6 +74,8 @@ public function get_css_wrapper_selector() {
 	}
 
 	public function save( $data ) {
+		$this->add_handle_revisions_changed();
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply core/kits/documents/kit.php]: Handled.

File: core/base/document.php [hunk 7]
```diff
@@ -33,6 +33,7 @@ abstract class Document extends Controls_Stack {
 	 */
 	const TYPE_META_KEY = '_elementor_template_type';
 	const PAGE_META_KEY = '_elementor_page_settings';
+	const CRC32_META_KEY = '_elementor_post_crc32';
```
Comment (matipojo) [inline core/base/document.php:36]: Remove.

File: core/base/document.php [hunk 8]
```diff
@@ -1,17 +1,17 @@
...
+use Elementor\Controls_Stack;
 use Elementor\Core\Files\CSS\Post as Post_CSS;
+use Elementor\Core\Settings\Manager as SettingsManager;
+use Elementor\Core\Settings\Page\Manager as PageManager;
 use Elementor\Core\Utils\Exceptions;
-use Elementor\Plugin;
 use Elementor\DB;
-use Elementor\Controls_Manager;
-use Elementor\Controls_Stack;
+use Elementor\Plugin;
```
Comment (matipojo) [inline core/base/document.php:11]: Cancel, leave it for a "Code Formatting" session.

File: core/base/document.php [hunk 9]
```diff
@@ -1284,4 +1289,29 @@ protected function get_post_statuses() {
...
+		// In case default, didn't determine the changes.
+		if ( ! $post_has_changed ) {
+			$last_revision_id = $last_revision->ID;
+			$last_revision_document = Plugin::instance()->documents->get( $last_revision_id );
+			$post_document = Plugin::instance()->documents->get( $post->ID );
+
+			$last_revision_settings = $last_revision_document->get_settings();
+			$post_settings = $post_document->get_settings();
+
+			// Its better to add crc32 signature for each revision and then only compare one part of the checksum.
```
Comment (matipojo) [inline core/base/document.php:1303]: `// TODO:`

File: tests/phpunit/elementor/core/kits/documents/kits/test-kit.php [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+		wp_set_current_user( $this->factory()->get_administrator_user()->ID );
+
+		$kit = Plugin::$instance->kits_manager->get_active_kit();
+		$this->kit = Plugin::$instance->documents->get( $kit->get_id(), false );
+
+		// In the production environment 'JS' sends empty array, do the same.
+		add_post_meta( $kit->get_main_id(), '_elementor_data', '[]' );
+	}
+
+	public function test_save_kits_revision_ensure_changed() {
```
Comment (matipojo) [inline tests/phpunit/elementor/core/kits/documents/kits/test-kit.php:25]: Separate the method name from the subject with `__`.
```suggestion
	public function test_save___kits_revision_ensure_changed() {
```

File: tests/phpunit/elementor/core/kits/documents/kits/test-kit.php [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+		$excepted_count++;
+
+		// Act.
+		$this->kit->save( [ 'settings' => $this->kit->get_settings() ] );
+
+		// Assert.
+		$this->assertCount( $excepted_count, Revisions_Manager::get_revisions( $this->kit->get_main_id() ) );
+	}
+
+	public function test_save_kits_revision_ensure_same() {
```
Comment (matipojo) [inline tests/phpunit/elementor/core/kits/documents/kits/test-kit.php:41]: As above.

File: tests/phpunit/elementor/core/kits/documents/kits/test-kit.php [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+		// Arrange.
+		$this->kit->set_settings( 'post_status', 'publish');
+		$this->kit->save( [ 'settings' => $this->kit->get_settings() ] );
+
+		$this->kit->set_settings( 'post_status', 'publish');
```
Comment (matipojo) [inline tests/phpunit/elementor/core/kits/documents/kits/test-kit.php:46]: Why?

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/base/document.php (hunk 9 above)
Comment (iNewLegend) [reply core/base/document.php]: Handled.

File: core/base/document.php (hunk 7 above)
Comment (iNewLegend) [reply core/base/document.php]: Handled.

File: core/base/document.php (hunk 8 above)
Comment (iNewLegend) [reply core/base/document.php]: Handled.

File: tests/phpunit/elementor/core/kits/documents/kits/test-kit.php (hunk 10 above)
Comment (iNewLegend) [reply tests/phpunit/elementor/core/kits/documents/kits/test-kit.php]: Handled.

File: tests/phpunit/elementor/core/kits/documents/kits/test-kit.php (hunk 11 above)
Comment (iNewLegend) [reply tests/phpunit/elementor/core/kits/documents/kits/test-kit.php]: Handled.

File: tests/phpunit/elementor/core/kits/documents/kits/test-kit.php (hunk 12 above)
Comment (iNewLegend) [reply tests/phpunit/elementor/core/kits/documents/kits/test-kit.php]: Handled.

File: tests/phpunit/elementor/core/kits/test-manager.php [hunk 13]
```diff
@@ -69,6 +66,9 @@ public function test_when_updating_blogname_or_blogdescription_option_it_should_
 		update_option( 'blogname', $expected_name );
 		update_option( 'blogdescription', $expected_description );
 
+		// Make sure that it fetched again from the DB and not from the cache.
```
Comment (matipojo) [inline tests/phpunit/elementor/core/kits/test-manager.php:69]: Why?

**Review by matipojo — CHANGES_REQUESTED** [review]

File: tests/phpunit/elementor/core/kits/test-manager.php (hunk 13 above)
Comment (iNewLegend) [reply tests/phpunit/elementor/core/kits/test-manager.php]: Handled.

**Review by matipojo — DISMISSED** [review]

File: core/base/document.php [hunk 14]
```diff
@@ -1284,4 +1288,29 @@ protected function get_post_statuses() {
...
+		return $post_has_changed;
+	}
+
+	public function add_handle_revisions_changed() {
+		add_filter( 'wp_save_post_revision_post_has_changed', [ $this, 'handle_revisions_changed' ], 10, 3 );
+	}
+
+	public function remove_handle_revisions_changed() {
+		remove_filter( 'wp_save_post_revision_post_has_changed', [ $this, 'handle_revisions_changed' ] );
+	}
```
Comment (kobizz) [inline core/base/document.php:1315]: 1. Make them private.
2. Consider naming the methods something like `<add|remove>_handle_revisions_post_has_changed_filter` for better clarifity.

**Review by kobizz — CHANGES_REQUESTED** [review]

File: core/base/document.php (hunk 14 above)
Comment (iNewLegend) [reply core/base/document.php]: Handled.

**Review by matipojo — APPROVED** [review]

**Review by kobizz — APPROVED** [review]

**Review by KingYes — APPROVED** [review]

**Pull request merged**